persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -0,0 +1,205 @@
+							func(tx persistence.ManagedTransaction) error {
+								err := tx.SaveOffset(
+									tc.Context,
+									"<source-app-key>",
+									conflictingCurrentOffset,
+									nextOffset,
+								)
+								gomega.Expect(err).To(gomega.Equal(offsetstore.ErrConflict))
+
+								return nil // let the transaction commit despite error
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+						actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+						gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
+					},
+					table.Entry("too low pair", 0, 1),
+					table.Entry("too high pair", 99, 100),
+				)
+			})
+
+			ginkgo.When("called multiple times within the same transaction", func() {
+				ginkgo.It("saves the last offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+
+					err := persistence.WithTransaction(
+						tc.Context,
+						dataStore,
+						func(tx persistence.ManagedTransaction) error {
+							err := tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							c = n
+							n += 123
+
+							err = tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							return nil
+						},
+					)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
+				})
+			})
+
+			ginkgo.When("the transaction is rolled-back", func() {
+				ginkgo.It("does not save the offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+
+					err := common.WithTransactionRollback(
+						tc.Context,
+						dataStore,
+						func(tx persistence.ManagedTransaction) error {
+							return tx.SaveOffset(
+								tc.Context,
+								"<source-app-key>",
+								c,
+								n,
+							)
+						},
+					)
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).To(gomega.BeEquivalentTo(c))
+				})
+			})
+
+			ginkgo.When("used in concurrent transactions", func() {
+				ginkgo.It("saves offsets concurrently", func() {
+					var g sync.WaitGroup
+
+					fn := func(ak string, c, n offsetstore.Offset) {
+						defer ginkgo.GinkgoRecover()
+						defer g.Done()
+
+						err := persistence.WithTransaction(
+							tc.Context,
+							dataStore,
+							func(tx persistence.ManagedTransaction) error {
+								return tx.SaveOffset(tc.Context, ak, c, n)
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					}
+
+					g.Add(3)
+					go fn("<source-app1-key>", 0, 1)
+					go fn("<source-app2-key>", 0, 2)
+					go fn("<source-app3-key>", 0, 3)
+					g.Wait()
```

Comment:
This test needs to assert that the expected offsets were actually stored against the right keys and all that, after the goroutines are finished.